Add a post: Fine-Tuning for Semantic Parsing with Levanter

Ivan-Zhou opened the pull request:
First PR of 2024! 

In this PR, I added a post on how to fine-tune Llama2-7B on GEM ViGGO using Levanter. I compared both full-weight fine-tuning and LoRA fine-tuning, with the out-of-box Llama2-7B Chat. 

Head:
<img width="1087" alt="image" src="https://github.com/stanford-crfm/levanter/assets/8097756/2f2cbf53-8957-4fa3-8b93-cd0b7049c8cd">

Results:
<img width="1092" alt="image" src="https://github.com/stanford-crfm/levanter/assets/8097756/66951622-1d2a-4009-9d70-30270047ad77">

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
````diff
@@ -0,0 +1,273 @@
+--hf_save_path gs://<somewhere>"
+```
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
````
🐎

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
+
+Below shows an example where we compare the output from the original chat model and the fine-tuned model. We can see that the fine-tuned model is able to generate the correct function call and precisely capture the right attributes, while the original model is not able to do so.
```
maybe give some numbers (baseline vs fine tuned) right before the example, just to set some expectations. Typical flow for something like this is `task overview -> dataset overview -> approach -> baseline results -> (exp -> results -> error analysis/example)*`

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
````diff
@@ -0,0 +1,273 @@
+
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
````
can you maybe format so that the outputs start at the same column so it's easy to visually see the comparison

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+
+Below is our configuration for LoRA fine-tuning. Note that it is very similar to the full-weight fine-tuning configuration, except for a few differences:
+
+- We increased the number of steps by 1 more epoch. LoRA fine-tuning is more efficient and less prone to overfitting, so we can train for more steps.
```
more efficient should be qualified a bit. My experience is it's slower but more ram and parameter efficient. lower capacity means it's regularized better

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+- We increased the number of steps by 1 more epoch. LoRA fine-tuning is more efficient and less prone to overfitting, so we can train for more steps.
+- We increased the learning rate to 3e-4, but we did not do very thorough hyperparameter tuning. We expect there might be a better learning rate.
+- We found weight decay at 0.1 to lead to better results than no weight decay, so we set it at 0.1.
+- We added the `lora` section to specify the LoRA parameters. All of the parameters are set to the default values.
```
maybe move this first?

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
````diff
@@ -0,0 +1,273 @@
+  dropout: 0.0  # dropout probability for LoRA layers
+```
+
+After training, Levanter will automatically merge the new parameters into the original model and save the new model. We can use it for inference and evaluation.
````
this depends on how it's invoked. with `hf_save_path` it will just save the lora adapter that works with PEFT (so just the lora weights plus a pointer), with merged_hf_save_path it will save a merged checkpoint.

dlwh replied:
it would be good to show the invocation for alpaca_lora because the new lora tutorial doesn't call alpaca_lora

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+## Evaluation
+
+### Metrics
+How do we accurately evaluate a model's performance in function calling tasks?
```
maybe move this earlier?

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+
+There are a few highlights from the results:
+
+- The baseline Llama2-7B Chat model's performance is remarkably low at 0.183 overall accuracy. This is consistent with our earlier observation of its limited capability on this task.
```
i'd say it doesn't really know what's expected? "capability" is a bit loaded.

dlwh commented on docs/examples/Fine-Tuning-Function-Calling.md:
```diff
@@ -0,0 +1,273 @@
+There are a few highlights from the results:
+
+- The baseline Llama2-7B Chat model's performance is remarkably low at 0.183 overall accuracy. This is consistent with our earlier observation of its limited capability on this task.
+- Fine-tuning methods, both full-weight and LoRA, substantially enhance the model's accuracy, achieving 0.780 and 0.748, respectively. Notably, LoRA fine-tuning approaches the performance of full-weight fine-tuning while adjusting only a fraction of the parameters, showcasing its efficiency.
```
might be good to say what the fraction is somewhere, (roughly)

dlwh commented on mkdocs.yml:
```diff
@@ -88,6 +88,7 @@ nav:
       - 'Configuration-Guide.md'
       - "Fine-Tuning.md"
       - "LoRA.md"
+      - "examples/Fine-Tuning-Function-Calling.md"
```
should we maybe break out an examples section? with Fine-Tuning, LoRA, and this?

Ivan-Zhou replied:
Yes, or `Tutorial` that you added would be great for this.

dlwh reviewed (commented):
thanks for putting this together. A few thoughts but otherwise i think this will be helpful

Ivan-Zhou commented on docs/examples/Fine-Tuning-Function-Calling.md:
````diff
@@ -0,0 +1,273 @@
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+Generation from Llama2-7B Chat: \nOutput: inform(name[SpellForce 3], release_year[2017], esrb[NA], genres[fantasy,\naction, adventure], platforms[PC], available_on_steam[no], has_linux_release[no],\nhas_mac_release[no])
````
It actually generate `\n`

Ivan-Zhou commented:
@dlwh Thanks for your comments!

Regarding your comment "is this a good example for function calling" -- yes I agree "semantic parsing" is a more appropriate term for this task, or we should call it "Functional Representation", which I believe is the most precise. 

When I was looking for a dataset, my primary intention was to look for an example of text to json (`Text -> Logical Form`  in your word). Essentially converting text to a structure format that can be parsed in Pydantic/Json. 

Later I found another dataset that is better to fit into the category of "Function Calling": https://huggingface.co/datasets/glaiveai/glaive-function-calling-v2. We can rename the current post as "Functional Representation", and then use the other dataset for another tutorial on Function Calling if needed.

dlwh commented:
Is functional representation a known technical term for this? I'd prefer to give it a known term of art?

Ivan-Zhou commented:
@dlwh I have updated the PR based on your comments. A few major changes:
1. I re-position this example from "Function Calling" to "Semantic Parsing". 
2. I re-picked all examples used to describe the task and error analysis; 
3. I added a launching script for LoRA;
4. I added more analysis on the confusion among function names by fine-tuned models;
5. I moved this post to the `Tutorial` section in mkdoc;

dlwh commented on docs/tutorials/Fine-Tuning-Semantic-Parsing.md:
```diff
@@ -0,0 +1,352 @@
+We evaluated the fine-tuned models on a hold-out evaluation set of 714 examples and computed the metrics described above.
+The results are shown in the table below.
+
+| Metric | Llama2-7B Chat | Full-weight Fine-tuning | LoRA Fine-tuning |
```
can you transpose this table? it's usually models as rows

dlwh commented on docs/tutorials/Fine-Tuning-Semantic-Parsing.md:
```diff
@@ -0,0 +1,352 @@
+
+- A notable area of confusion is within the `confirm` column, where both models frequently misclassify examples from the `inform`, `recommend`, and `suggest` categories as `confirm`. This may stem from these function names being too similar in the dataset, leading to ambiguity that the models struggle to differentiate.
+- The full-weight fine-tuning model (depicted on the left) tends to make more errors in the "empty" or "other" columns, indicating instances where the model either fails to predict a function name or predicts one that doesn't exist within the predefined set.
+- The LoRA fine-tuning model (shown on the right) exhibits more confusion between the predefined function classes; overall, it has fewer correct predictions along the diagonal of the matrix.
```
this is weird to me because it goes against the narrative in my head that LoRA is good for teaching the structure of a task but bad for injecting new information. Do you have any thoughts?

Ivan-Zhou replied:
Yes, the amount of confusion is indeed, bit high. 

I think the results on attributes fields and value does show the advantage of fine-tuning at "teaching about form/structure": both models outputs attributes in the correct format and contains the correct information. 

What I want to show here is that, the selection of the function names is more ambiguous and difficult. It is beyond the "form" problem, but more on reasoning, and the difference between functions `confirm`, `inform`, `suggest` and `recommend` does not seem clear. So I think reducing # functions, clarify the difference, and possibly improve label quality, would be key to improve on the function name accuracy.

dlwh reviewed (commented):
sorry i was slow, a few more nits

Ivan-Zhou commented on docs/tutorials/Fine-Tuning-Semantic-Parsing.md:
```diff
@@ -0,0 +1,352 @@
+# Fine-Tuning for Semantic Parsing with Levanter
+
+To enable computers to assist human with complex tasks, such as data anlaysis or customer service, they must first understand our natural language and convert it into a format that they can process.
```
I re-wrote the introduction to be more specific on semantic parsing. Hope this version would be more accurate and proper!

dlwh commented on docs/tutorials/Fine-Tuning-Semantic-Parsing.md:
````diff
@@ -350,3 +347,81 @@ In this post, we showcase the process of fine-tuning a Llama2 model using Levant
+### Appendix A: The Prompt Used in This Task
+
+```
+You are a helpful chatbot to assist users to parse queries in natural language into structural format.
````
oh man this is long. I think it's not necessary to use such a long prompt for finetuning (especially I don't think you need few shot examples with fine tuning?)

dlwh reviewed (approved):
lgtm!
